No longer able to place types in top-level modules

### What happened?

There appears to be a regression in the ability to place types in external modules.

To reproduce, assign a following token to a resource:

```
		Resources: map[string]*tfbridge.ResourceInfo{
			"myprovider_myresource": {
				Tok: "myprovider:mymodule:MyResource",
			},
		},
```

This is not a typical token allocation strategy as most Pulumi providers use  "myprovider:myparentmodule/mymodule:MyResource", however it is used in the wild and the ability to do this has regressed, causing upgrade issues.

Current behavior during `make tfgen`:

```
panic: fatal: An assertion has failed: QName "myModule" has invalid namespace ""

goroutine 1 [running]:
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
        /Users/anton.tayanovskyy/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.75.0/go/common/util/contract/failfast.go:23
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Assertf(0x0?, {0x1023fd4f4?, 0x1028ec168?}, {0x140008bb2e8?, 0x1400030ba20?, 0x15?})
        /Users/anton.tayanovskyy/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.75.0/go/common/util/contract/assert.go:35 +0xe0
github.com/pulumi/pulumi/sdk/v3/go/common/tokens.QName.Namespace({0x1400030ba2b, 0xa})
        /Users/anton.tayanovskyy/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.75.0/go/common/tokens/names.go:117 +0x110
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen.parentModuleName(...)
        /Users/anton.tayanovskyy/code/pulumi-terraform-bridge/pkg/tfgen/generate_schema.go:1041
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen.parentModule({0x1400030ba20, 0x15})
        /Users/anton.tayanovskyy/code/pulumi-terraform-bridge/pkg/tfgen/generate_schema.go:1036 +0x3c
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen.parentModuleOrDefault(...)
        /Users/anton.tayanovskyy/code/pulumi-terraform-bridge/pkg/tfgen/generate_schema.go:1026
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen.modulePlacementForType({0x14000a8af50, 0xa}, {0x102f124f0, 0x1400097d3b0})
        /Users/anton.tayanovskyy/code/pulumi-terraform-bridge/pkg/tfgen/generate_schema.go:1011 +0x1fc
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen.modulePlacementForTypeSet({0x14000a8af50, 0xa}, 0x14000304c40?)
        /Users/anton.tayanovskyy/code/pulumi-terraform-bridge/pkg/tfgen/generate_schema.go:971 +0x11c
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen.(*schemaGenerator).genObjectTypeToken(0x140008bd8f8, 0x14000465480)
        /Users/anton.tayanovskyy/code/pulumi-terraform-bridge/pkg/tfgen/generate_schema.go:650 +0x94
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen.(*schemaGenerator).genPackageSpec(_, _)
        /Users/anton.tayanovskyy/code/pulumi-terraform-bridge/pkg/tfgen/generate_schema.go:244 +0x1d78
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen.genPulumiSchema(_, {_, _}, {_, _}, {{0x102f26300, 0x14000a9c380}, {0x1023c144a, 0xa}, {0x0, ...}, ...}, ...)
        /Users/anton.tayanovskyy/code/pulumi-terraform-bridge/pkg/tfgen/generate_schema.go:209 +0xbc
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen.(*Generator).Generate(0x14000788900)
        /Users/anton.tayanovskyy/code/pulumi-terraform-bridge/pkg/tfgen/generate.go:856 +0xdc
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen.Main.func1({{0x1023c144a, 0xa}, {0x1028ec180, 0x9}, {0x16f0ef8eb, 0x6}, {{0x102f26300, 0x14000a9c380}, {0x1023c144a, 0xa}, ...}, ...})
        /Users/anton.tayanovskyy/code/pulumi-terraform-bridge/pkg/tfgen/main.go:44 +0x50
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen.newTFGenCmd.func1(0x0?, {0x14000a45110, 0x1, 0x0?})
        /Users/anton.tayanovskyy/code/pulumi-terraform-bridge/pkg/tfgen/main.go:162 +0x498
github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil.RunFunc.func1(0x0?, {0x14000a45110?, 0x0?, 0x0?})
        /Users/anton.tayanovskyy/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.75.0/go/common/util/cmdutil/exit.go:96 +0x28
github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil.RunResultFunc.func1(0x14000788300?, {0x14000a45110, 0x1, 0x3})
        /Users/anton.tayanovskyy/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.75.0/go/common/util/cmdutil/exit.go:112 +0x38
github.com/spf13/cobra.(*Command).execute(0x14000788300, {0x1400004c090, 0x3, 0x3})
        /Users/anton.tayanovskyy/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944 +0x5ac
github.com/spf13/cobra.(*Command).ExecuteC(0x14000788300)
        /Users/anton.tayanovskyy/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x340
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/anton.tayanovskyy/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen.MainWithCustomGenerate({_, _}, {_, _}, {{0x102f26300, 0x14000a9c380}, {0x1023c144a, 0xa}, {0x0, 0x0}, ...}, ...)
        /Users/anton.tayanovskyy/code/pulumi-terraform-bridge/pkg/tfgen/main.go:57 +0x54
github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfgen.Main({_, _}, {_, _}, {{0x102f26300, 0x14000a9c380}, {0x1023c144a, 0xa}, {0x0, 0x0}, ...})
        /Users/anton.tayanovskyy/code/pulumi-terraform-bridge/pkg/tfgen/main.go:36 +0x58

```

### Expected Behavior

"myprovider:mymodule:MyResource" works as before.

### Steps to reproduce

make tfgen on top of

```
Resources: map[string]*tfbridge.ResourceInfo{
			"myprovider_myresource": {
				Tok: "myprovider:mymodule:MyResource",
			},
		},
```

### Output of `pulumi about`

N/A

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/pkg/tfgen/generate_schema.go
+++ b/pkg/tfgen/generate_schema.go
@@ -1002,14 +1002,15 @@ func modulePlacementForType(pkg tokens.Package, path paths.TypePath) tokens.Modu
 			// root module).
 			return res.Token().Module()
 		}
-		// Supplementary types are defined one level up from
-		// the module defining the resource.
-		return parentModule(res.Token().Module())
+		// Supplementary types are typically defined one level up from the module defining the resource, but may
+		// also be defined in the same module.
+		m := res.Token().Module()
+		return parentModuleOrDefault(m, m)
 	case *paths.DataSourceMemberPath:
-		dataSourceModule := pp.DataSourcePath.Token().Module()
-		// Supplementary types are defined one level up from
-		// the module defining the data source.
-		return parentModule(dataSourceModule)
+		// Supplementary types are typically defined one level up from the module defining the data source, but
+		// may also be defined in the same module.
+		m := pp.DataSourcePath.Token().Module()
+		return parentModuleOrDefault(m, m)
 	case *paths.ConfigPath:
 		return tokens.NewModuleToken(pkg, configMod)
 	default:
@@ -1018,8 +1019,18 @@ func modulePlacementForType(pkg tokens.Package, path paths.TypePath) tokens.Modu
 	}
 }
 
-func parentModule(m tokens.Module) tokens.Module {
-	return tokens.NewModuleToken(m.Package(), parentModuleName(m.Name()))
+func parentModuleOrDefault(m tokens.Module, def tokens.Module) tokens.Module {
+	if m, ok := parentModule(m); ok {
+		return m
+	}
+	return def
+}
+
+func parentModule(m tokens.Module) (tokens.Module, bool) {
+	if !strings.Contains(string(m.Name()), tokens.QNameDelimiter) {
+		return "", false
+	}
+	return tokens.NewModuleToken(m.Package(), parentModuleName(m.Name())), true
 }
 
 func parentModuleName(m tokens.ModuleName) tokens.ModuleName {

--- a/pkg/tfgen/generate_schema.go
+++ b/pkg/tfgen/generate_schema.go
@@ -1005,12 +1005,12 @@ func modulePlacementForType(pkg tokens.Package, path paths.TypePath) tokens.Modu
 		// Supplementary types are typically defined one level up from the module defining the resource, but may
 		// also be defined in the same module.
 		m := res.Token().Module()
-		return parentModuleOrDefault(m, m)
+		return parentModuleOrSelf(m)
 	case *paths.DataSourceMemberPath:
 		// Supplementary types are typically defined one level up from the module defining the data source, but
 		// may also be defined in the same module.
 		m := pp.DataSourcePath.Token().Module()
-		return parentModuleOrDefault(m, m)
+		return parentModuleOrSelf(m)
 	case *paths.ConfigPath:
 		return tokens.NewModuleToken(pkg, configMod)
 	default:
@@ -1019,11 +1019,11 @@ func modulePlacementForType(pkg tokens.Package, path paths.TypePath) tokens.Modu
 	}
 }
 
-func parentModuleOrDefault(m tokens.Module, def tokens.Module) tokens.Module {
-	if m, ok := parentModule(m); ok {
+func parentModuleOrSelf(self tokens.Module) tokens.Module {
+	if m, ok := parentModule(self); ok {
 		return m
 	}
-	return def
+	return self
 }
 
 func parentModule(m tokens.Module) (tokens.Module, bool) {

--- a/pkg/tfgen/generate_test.go
+++ b/pkg/tfgen/generate_test.go
@@ -15,6 +15,7 @@
 package tfgen
 
 import (
+	"fmt"
 	"io"
 	"os"
 	"path/filepath"
@@ -280,4 +281,86 @@ func Test_ProviderWithObjectTypesInConfigCanGenerateRenames(t *testing.T) {
 	})
 	require.NoError(t, err)
 	assert.Equal(t, "foo_bar", r.Renames.RenamedProperties["test:index/ProviderProp:ProviderProp"]["fooBar"])
+}
+
+func TestModulePlacementForType(t *testing.T) {
+	t.Parallel()
+
+	type testCase struct {
+		pkg    tokens.Package
+		path   paths.TypePath
+		expect tokens.Module
+	}
+
+	prop := paths.PropertyName{Key: "prop", Name: "prop"}
+
+	testCases := []testCase{
+		// Resource in top-level module mymod is placed to myprov:mymod.
+		{
+			"myprov",
+			paths.NewProperyPath(paths.NewResourcePath(
+				"myprov_myres",
+				"myprov:mymod:MyRes",
+				false, /*isProvider*/
+			).Inputs(), prop),
+			"myprov:mymod",
+		},
+		// Resource in second-level module mymod is placed to parent myprov:mymod.
+		{
+			"myprov",
+			paths.NewProperyPath(paths.NewResourcePath(
+				"myprov_myres",
+				"myprov:mymod/mysubmod:MyRes",
+				false, /*isProvider*/
+			).Inputs(), prop),
+			"myprov:mymod",
+		},
+		// Resource in third-level module mymod is placed to parent myprov:mymod/mysubmod.
+		{
+			"myprov",
+			paths.NewProperyPath(paths.NewResourcePath(
+				"myprov_myres",
+				"myprov:mymod/mysubmod/mysubsubmod:MyRes",
+				false, /*isProvider*/
+			).Inputs(), prop),
+			"myprov:mymod/mysubmod",
+		},
+		// Datasource in top-level module mymod is placed to myprov:mymod.
+		{
+			"myprov",
+			paths.NewProperyPath(paths.NewDataSourcePath(
+				"myprov_myds",
+				"myprov:mymod:MyFn",
+			).Args(), prop),
+			"myprov:mymod",
+		},
+		// Datasource in second-level module mymod is placed to parent myprov:mymod.
+		{
+			"myprov",
+			paths.NewProperyPath(paths.NewDataSourcePath(
+				"myprov_myds",
+				"myprov:mymod/mysubmod:MyFn",
+			).Args(), prop),
+			"myprov:mymod",
+		},
+		// Datasource in third-level module mymod is placed to parent myprov:mymod/mysubmod.
+		{
+			"myprov",
+			paths.NewProperyPath(paths.NewDataSourcePath(
+				"myprov_ds",
+				"myprov:mymod/mysubmod/mysubsubmod:MyFn",
+			).Args(), prop),
+			"myprov:mymod/mysubmod",
+		},
+	}
+
+	for i, tc := range testCases {
+		tc := tc
+		t.Run(fmt.Sprintf("testCase-%d", i), func(t *testing.T) {
+			t.Parallel()
+			mod := modulePlacementForType(tc.pkg, tc.path)
+			assert.Equal(t, tc.expect, mod)
+		})
+	}
+
 }

